File: LeetCode/cpp/141.cpp
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(!head || !head->next)
            return false;
        auto slow = head->next, fast = head->next->next;
        while (fast && fast->next && fast != slow) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow == fast;
    }
};

File: LeetCode/python/141.py
class Solution(object):
    def hasCycle(self, head):
        p, q = head, head
        while True:
            if p and p.next:
                p = p.next.next
            else:
                return False
            q = q.next
            if q == p:
                return True
        return False
